subt/simulation.py

```diff
@@ -154,6 +157,53 @@ def on_desired_speed(self, dt, channel, data):
         self.bus.publish('rot', [heading_centidegrees, 0, 0])
         self.bus.publish('pose2d', [x_mm, y_mm, heading_centidegrees])
         self.bus.publish('sim_time_sec', self.time.total_seconds())
-        self.bus.publish('scan', [0] * 100)
+        self.bus.publish('scan', self.scan())
 
         #log("++++", x, y, math.degrees(heading))
+
+    def scan(self):
+        from shapely.affinity import translate, rotate
+        "270 degrees -> np.linspace(-135, 135, 271)"
+        ret = np.zeros((271,))
+
+        # create lidar lines
+        angles = np.linspace(math.radians(-135), math.radians(135), 271)
+        acos = np.cos(angles)
+        asin = np.sin(angles)
+        max_dist_m = 10.0
+        x = acos * max_dist_m
+        y = asin * max_dist_m
+
+        # transform world to robot coordinates
+        world = self.world
+        world = translate(world, *[-c for c in self.xyz])
+        world = rotate(world, -quaternion.heading(self.orientation), origin=(0,0), use_radians=True)
+
+        # intersect each lidar line with the world
+        for i, point in enumerate(zip(x, y)):
+            cross = world.intersection(LineString([(0,0), point]))
+            if cross.is_empty:
+                continue
+            if cross.geom_type == 'Point':
+                ret[i] = math.sqrt(sum(c**2 for c in cross.coords[0]))
```

Comment:
math.hypot(*cross.coords) should work, no?

Reply:
right, I had that nagging feeling of something not being right when I wrote it :+1: